diff --git a/target/mips/tcg/translate.c b/target/mips/tcg/translate.c
index 0000000..0000000 100644
--- a/target/mips/tcg/translate.c
+++ b/target/mips/tcg/translate.c
@@ -14196,9 +14196,9 @@ static void decode_opc(CPUMIPSState *env, DisasContext *ctx)
     case OPC_BSHFL:
         op2 = MASK_BSHFL(ctx->opcode);
         switch (op2) {
         case OPC_ALIGN:
         case OPC_ALIGN_1:
         case OPC_ALIGN_2:
         case OPC_ALIGN_3:
-            check_insn(ctx, ISA_MIPS_R6);
-            decode_opc_special3_r6(env, ctx);
+            check_insn(ctx, ISA_MIPS_R2); /* SOC_REF_ALIGN_R2 */
+            gen_align(ctx, 32, rd, rs, rt, sa & 3);
             break;
